### PR TITLE
Deterministic PII Pseudonymization

### PR DESCRIPTION
**Pull request by Vidushi2709**
# Pull Request: Deterministic PII Pseudonymization

## Description

This PR implements deterministic PII pseudonymization for the Noveum Trace SDK. When enabled, sensitive personally identifiable information (emails, phone numbers, SSNs, credit cards, IPs, URLs, and NER-detected entities) is replaced with HMAC-SHA256-based deterministic pseudonyms before sending traces to the backend.

**Key features:**
- Opt-in via `NOVEUM_PII_ENABLED` environment variable (defaults to `false`)
- Configurable salt via `NOVEUM_PII_SALT` (defaults to `"noveum-pii-default-salt"`)
- Hybrid detection: spaCy NER (PERSON, GPE, ORG, LOC) + regex patterns (EMAIL, PHONE, SSN, CARD, IP, URL)
- Deterministic tokens allow correlation across traces with the same salt
- Dev mode keeps raw traces on disk for debugging
- No new hard dependencies (spaCy optional, gracefully falls back to regex-only mode)

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update
- [x] Code refactoring
- [x] Test improvements

## How Has This Been Tested?

- [x] Unit tests
- [ ] Integration tests
- [x] Manual testing
- [ ] Performance testing

### Test Coverage

**Unit Tests (16/16 PASSED):**
- test_pii_pseudonymizer.py (12 tests)
  - Deterministic token generation
  - Salt-dependent output
  - NFC normalization equivalence
  - Overlapping span deduplication (longest wins)
  - Email, phone, SSN, card, IP, URL regex detection
  - Right-to-left index preservation
  - Nested dict/list recursion
  - spaCy graceful fallback on ImportError/OSError

- test_config_comprehensive.py (1 test)
  - `NOVEUM_PII_ENABLED` environment variable parsing
  - `NOVEUM_PII_SALT` environment variable loading
  - Default salt application when not set
  - Config `to_dict()`/`from_dict()` round-trip

- test_http_transport_additional.py (3 tests)
  - Dev trace file written with raw data
  - POST request sent with pseudonymized payload
  - Batch operations properly pseudonymized
  - Disabled PII mode sends original payload

### Manual Testing

```bash
# Run all PII tests
python -m pytest tests/unit/utils/test_pii_pseudonymizer.py \
  tests/unit/core/test_config_comprehensive.py::TestConfigurationLoading::test_load_from_environment_pii_settings \
  tests/unit/transport/test_http_transport_additional.py -k "pii" -v

# Enable and test with env vars
export NOVEUM_PII_ENABLED=true
export NOVEUM_PII_SALT="custom-salt-123"
python your_script.py  # traces will be pseudonymized
```

## Test Configuration

* Python version: 3.11.9
* Operating System: Windows 11
* Dependencies:
  - `noveum-trace` (core)
  - `spacy` (optional, for NER)
  - `pytest` (testing)
  - `requests` (HTTP transport)

## Checklist

- [x] My code follows the style guidelines of this project
  - Passes `black`, `isort`, `ruff`, `mypy`, `bandit` checks
  - Pre-commit hooks pass on all files
  
- [x] I have performed a self-review of my own code
  - Code review completed; logic verified against spec
  
- [x] I have commented my code, particularly in hard-to-understand areas
  - Comprehensive docstrings on `PiiPseudonymizer` class and all methods
  - Inline comments on complex span deduplication logic
  
- [x] I have made corresponding changes to the documentation
  - Implementation follows existing patterns in AGENTS.md
  - Config changes documented inline with defaults
  
- [x] My changes generate no new warnings
  - All linters pass; no mypy errors; no type issues
  
- [x] I have added tests that prove my fix is effective or that my feature works
  - 16 comprehensive unit tests covering all code paths
  - 100% pass rate with edge case coverage
  
- [x] New and existing unit tests pass locally with my changes
  - All 16 PII tests pass
  - No regression in existing test suite
  
- [x] Any dependent changes have been merged and published in downstream modules
  - Feature is self-contained; no external dependencies

## Files Modified

1. **config.py**
   - Added `pii_enabled: bool = False` to `SecurityConfig`
   - Added `pii_salt: str = DEFAULT_PII_SALT` to `SecurityConfig`
   - Wired `NOVEUM_PII_ENABLED` env var loading (line 534)
   - Wired `NOVEUM_PII_SALT` env var loading (line 544)
   - Updated `to_dict()` and `from_dict()` for serialization

2. **pii_redaction.py**
   - Added `PiiPseudonymizer` class (line 258)
   - Implements deterministic HMAC-SHA256 pseudonymization
   - Hybrid NER + regex span detection
   - Deduplicates overlapping spans (longest wins)
   - Right-to-left replacement to preserve indices

3. **http_transport.py**
   - Instantiate `PiiPseudonymizer` in `__init__()` when `pii_enabled=True`
   - Apply pseudonymization in `_send_request()` after dev write
   - Apply pseudonymization in `_send_trace_batch()` after dev write

4. **Test files**
   - test_pii_pseudonymizer.py - 12 comprehensive tests
   - test_config_comprehensive.py - env var loading test
   - test_http_transport_additional.py - 3 integration tests

## Performance Impact

- **Minimal overhead**: O(n) single-pass through trace data
- **Optional**: Only active when explicitly enabled
- **Lazy load**: spaCy model loaded once per transport instance
- **Graceful degradation**: Falls back to regex-only if spaCy unavailable

## Backward Compatibility

✅ **Fully backward compatible**
- Feature is opt-in (defaults to `false`)
- No breaking changes to public APIs
- Existing code unaffected; no behavior change without explicit config

## Summary by CodeRabbit

* **New Features**
  * Added PII pseudonymization for traces (regex-based and optional NER) producing deterministic, salt-based tokens.
  * Configuration supports enabling PII protection and supplying a salt (env var support). Merging preserves file-provided nested security values when only env flags are set.
  * Transport sends pseudonymized trace payloads when PII protection is enabled; dev-mode on-disk traces remain raw.

* **Tests**
  * Added comprehensive tests for config, transport, and pseudonymizer behavior (including NER and edge cases).

### CODE REVIEW

File: src/noveum_trace/utils/pii_redaction.py [hunk 1]
```diff
@@ -250,3 +254,127 @@ def create_redaction_summary(original_text: str, redacted_text: str) -> dict[str
...
+
+
+class PiiPseudonymizer:
+    """
+    Deterministic pseudonymization of PII-like spans using HMAC-SHA256 + salt.
+
+    Uses spaCy ``en_core_web_sm`` when importable and the model is available;
+    otherwise relies on regex spans only. No mandatory dependency beyond the
+    standard library.
+    """
```
Comment (shivamgcodes) [inline src/noveum_trace/utils/pii_redaction.py:136]: did you need to add any libraries for the spaCy model?? can you please add it to pyproject.toml ?

File: src/noveum_trace/utils/pii_redaction.py [hunk 2]
```diff
@@ -5,7 +5,11 @@
 identifiable information from trace data.
 """
 
+import hashlib
+import hmac
+import importlib
 import re
+import unicodedata
```
Comment (shivamgcodes) [inline src/noveum_trace/utils/pii_redaction.py:14]: Kindly check thoroughly, if the rest of the pre-existing code in this file is useful / relevant. Please remove redundant / inefficient code if any.

**Review by shivamgcodes — APPROVED** [review]
